### ISSUE TITLE
Optimizer step outside gradient_accumulation condition in tutorial 4.1?

### ISSUE THREAD

**Issue by guoxuxu** [thunlp/OpenPrompt#106]
Hi, Thanks a lot for the tutorials. Hope to see more useful tutorials with OpenPrompt.
I feel that the following code in tutorial 4.1 has some problems:
```
loss.backward()
tot_loss += loss.item()
actual_step += 1

if actual_step % gradient_accumulation_steps == 0:
    torch.nn.utils.clip_grad_norm_(prompt_model.parameters(), 1.0)
    glb_step += 1
    if glb_step % pbar_update_freq == 0:
        ......
if optimizer1 is not None:
    optimizer1.step()
    optimizer1.zero_grad()
if scheduler1 is not None:
    scheduler1.step()
......
```
Refer to the LM-BFF [Code](https://github.com/princeton-nlp/LM-BFF/blob/main/src/trainer.py) line 350-378, which also used the gradient accumulation steps maybe to reduce the GPU overhead? The update of the optimizer actually should be written within the if-else loop, intuitively. 
```
loss.backward()
tot_loss += loss.item()
actual_step += 1

if actual_step % gradient_accumulation_steps == 0:
    torch.nn.utils.clip_grad_norm_(prompt_model.parameters(), 1.0)
    glb_step += 1
    if glb_step % pbar_update_freq == 0:
        ......
    if optimizer1 is not None:
        optimizer1.step()
        optimizer1.zero_grad()
    if scheduler1 is not None:
        scheduler1.step()
```

**Comment by ShengdingHu** [comment]
OMG, thanks for discovering the ridiculous bug we made, we are re-running the experiments. Thanks!
